lib/blockchain/transaction.rs

```diff
@@ -20,30 +20,25 @@ impl Transaction {
   /// Give each transaction a hash to be later used in merkle tree.
   pub fn transaction_hash(&self) -> String {
     let mut hasher = Sha256::new();
-    let total_sum: u64 = self.input_value() + self.output_value();
+    
+    self.input_hashes().iter().for_each(|hash| hasher.input(hash.as_bytes()));
+    self.output_hashes().iter().for_each(|hash| hasher.input(hash.as_bytes()));
 
-    hasher.input(total_sum.to_string().as_bytes());
     hasher.result_str()
   }
 
-  /// Hash the vector of output data by summing the bytes.
-  fn generate_hash(&self, to_hash: Vec<u8>) -> String {
+  /// Hash of output data.
+  fn generate_hash(&self, to_hash: &Output) -> String {
     let mut hasher = Sha256::new();
-    let to_hash_sum: u8 = to_hash.iter().sum();
     
-    hasher.input(to_hash_sum.to_string().as_bytes());
+    let output_value_string = to_hash.value.to_string();
+    let owned_address = to_hash.to_address.to_owned();
+    let combined_data = owned_address + &output_value_string;
+    
+    hasher.input(combined_data.as_bytes());
     hasher.result_str()
   }
 
-  /// Turn output data into vector of bytes to later be hashed.
-  fn output_bytes(&self, output: &Output) -> Vec<u8> {
-    let mut bytes = vec![];
-    bytes.extend(output.to_address.as_bytes());
-    bytes.extend(output.value.to_string().as_bytes());
-
-    bytes
-  }
-
   // Sum of value of each input.
   pub fn input_value(&self) -> u64 {
```

Comment:
This method (and `output_value`) don't really tell what they do from a quick read. Obviously looking at the signature and the code its obvious, but the fact that it needs a comment indicates a smell.

A small rename could help tremendously, maybe `fn get_sum()`.

A second more important point about this is the fact that both `input/output_value` have the same code and refer to an alien object is a smell; it seems like that function would be much better suited for a `impl Output` as:

```
impl Output {
    pub fn get_sum_of_values_from_vec(input: &Vec<Self>) -> u64 {
        input.iter().map(|output| output.value).sum()
    }
}
```

This would let us get rid of those ugly functions and replace them with nicer calls as well as giving us some room for abstraction if we ever decide to change the `Output` type.

Reply:
I merged these two methods into one but I left it in the Transaction type since it's summing transaction data, which is part of its struct. This way we can find the sum of either input or output using the Transaction object (I created a test showing this).